### ISSUE TITLE
RFC: Naming of the different Zephyr OS specification levels

### ISSUE THREAD

**Issue by romkell** [zephyrproject-rtos/reqmgmt#23]
## Current situation
The currently used names
- "high level specification" and
- "functional specification"
 
for the Zephyr OS specifications do not seem to be ideal to reflex their content. "High level" is not very specific and "functional" is often used in the context of "functional and non-functional".
They also do not seem to be near any name used by the various safety standards such as 61508, 26262, 50128, DO-178 and alike.
For easier reuse of downstream Zephyr OS user building systems which will comprise Zephyr OS and shall comply with the listed standards, more common names are an advantage.

## Goal

Find specification names that will be recognized in the respective industries because the name is close enough to their (safety, security?) standards.

## Approach

* Collect as many as possible specification names from the various standards in this RFC
* From that collection, conclude on the names of the Zephyr OS specifications

## Definitions
* Document:
  * Aome kind of container holding and presenting text artifacts in a structured manner, e.g. MS Word, strictdoc, Sphinx etc.
* Specification:
  * A collection of definitions, describing a system, software, architecture or component or alike depending on the level. A specification may but does not need to live in one document. It can be distributed among several documents, different levels of specification can be organized in the same document.

### IEC EN 61508-3:2010 [Industry in general]

- (E/E/PE safety requirements specification - as from 61508-1)
- (E/E/PE system safety requirements specification - as from 61508-2)
  - software safety requirements specification
  - software architecture design
  - software architecture integration test specification
  - software/ PE integration test specification
      - Software system design specification
      - software system integration test specification.
        - software module design specification
        - software module test specification

### EN 50128:2011 / EN 50657:2017 [Railway]

- (System Requirements Specification)
- (System Safety Requirements Specification - no longer in 50657)
- (System Architecture Description)
  - Software Requirements Specification
  - Overall Software Test Specification
    - Software Architecture Specification
    - Software Design Specification
    - Software Interface Specification
    - Software Integration Test Specification
    - Software/Hardware Integration Test Specification
      - Software Component Design Specification
      - Software Component Test Specification

### ISO 13849-1:2023 [Safety of machinery parts]

- Safety requirements specification

### IEC 62061:2021 [Safety of machinery]

- software safety requirements specification - SW level 1, 2
- software design specification - SW level 1, 2
- software system design specification - SW level 2
- Module design specification - SW level 1, 2
- Test specs are not explicitly named 

### EN 60601 [Medical]

- tbd

### ISO 26262 [Automotive]

- Software Safety Requirements Specification
- Software Architectural Design Specification (iso26262-6.7)
- Software Unit Design
- Software Verification Specification
- Software Verification Report

### DO-178 [Airborne]

- tbd

**Comment by quarterbit** [comment]
For ISO 26262 [Automotive] some of the relevant namings:
- Software Safety Requirements Specification 
- Software Architectural Design Specification (iso26262-6.7) 
- Software Unit Design
- Software Verification Specification 
- Software Verification Report

As a starting point.

**Comment by romkell** [comment]
> For ISO 26262 [Automotive] some of the relevant namings:
> 
>     * Software Safety Requirements Specification
> 
>     * Software Architectural Design Specification (iso26262-6.7)
> 
>     * Software Unit Design
> 
>     * Software Verification Specification
> 
>     * Software Verification Report
> 
> 
> As a starting point.

Cheers.

**Comment by legrand-gregshue** [comment]
Architecting/designing a solution is a recursive decomposition activity. The elements identified in one level get broken down into sub-elements in the next level. This recurses as many times as necessary to reach sub-elements with low enough complexity to be directly implemented and verified. This activity starts with solution providers/product manufacturers. They may have multiple recursions before choosing to introduce the Zephyr ecosystem and may need multiple recursions for proprietary subsystems/drivers they introduce. (IIRC, NASA publicly discloses they have typically had 5-8 recursion levels in their system designs.)

What is the plan for identifying/organizing/relating these documents when multiple recursion levels exist?

**Comment by stanislaw** [comment]
Space industry:

**TL;DR**

Customer-level:

- Statement of Work 
- (One or more) Technical Requirements Specifications (spacecraft subsystems, such as power, thermal, ..., including software subsystems)
- (One or more) Software System Specification + Interface Requirements Document 

All of these documents typically contains inputs from other guidelines and standards and make them directly applicable to what supplier is to produce.

Supplier-level:

- Software Requirements Specification + Interface Control Document
- Software Design Document
- Software User Manual
- many other such as Software Reuse File for dealing with third-party SW or SW FMEA, ...

---

Here's a longer version with a disclaimer: I didn't see / attend to the whole world of spacecraft development :)

The European space industry's documents governed by the ECSS standards assume a customer-supplier relationship.

The customer-level software requirements document is typically called **Software System Specification (SSS)**. The supplier-level – **Software Requirements Specification (SRS)**.
The SSS focuses on the WHAT of a to-be-developed product and is written by a customer.
The SRS focuses on the HOW and is written by a supplier.

This highly depends on the contractual arrangements and the types of the companies, but very often a customer and a supplier DO NOT exchange the SSS and SRS documents directly.
Instead, a customer provides an **Interface Requirements Document (IRD)** to a supplier,
and the supplier answers with an **Interface Control Document (ICD)**. The communication therefore happens only at the black-box interface level. Whether or not the customer sees into the details of the supplier's again depends on both companies' experience and a contractual arrangement.

If a project is small enough or there is a full transparency, the IRD and ICD could be merged back to the SSS and SRS respectively and exchanged directly between a customer and a supplier. In that case, they simply become chapters of "Interface Requirements" in the SSS and SRS documents.

Finally, in accordance with @legrand-gregshue's comment just above, the SSS is typically not a top-level document in a specification tree. If a project is initiated by the European Space Agency,
it can start with a Statement of Work document that contains the most high-level requirements.
Depending on the size of a program, the next levels can include one or more **Technical Requirements Specifications** which cover and breakdown into all spacecraft subsystems, including software. These documents can receive inputs from a number of other documents and agency/company-level guidelines to completely represent a solid background for all inputs to the SSS. Only starting from this level, the actual Software System Specification starts to be written. I have not yet seen an SSS to be a top-level document in a space project.

Zephyr is not a spacecraft but I would still assume that the WHAT and HOW requirements levels must be well separated as if the Zephyr development implemented a "safe and secure RTOS" (WHAT). This is what I also do for StrictDoc and it seems to work well as in the "A requirements tool" vs "The StrictDoc requirements tool" levels.

**Comment by kestewart** [comment]
From discussions in the Safety working group this has been decided for the project and merged in.   This should satisfy 
https://github.com/zephyrproject-rtos/zephyr/pull/71953
